planner/core/stats.go

```diff
@@ -554,8 +557,15 @@ func (ds *DataSource) buildIndexMergeOrPath(partialPaths []*util.AccessPath, cur
 	indexMergePath := &util.AccessPath{PartialIndexPaths: partialPaths}
 	indexMergePath.TableFilters = append(indexMergePath.TableFilters, ds.pushedDownConds[:current]...)
 	indexMergePath.TableFilters = append(indexMergePath.TableFilters, ds.pushedDownConds[current+1:]...)
+	tableFilterCnt := 0
 	for _, path := range partialPaths {
+		// IndexMerge should not be used when the SQL is like 'select x from t WHERE (key1=1 AND key2=2) OR (key1=4 AND key3=6);'.
+		// Check issue https://github.com/pingcap/tidb/issues/22105 for details.
```

Comment:
Actually we can use IndexMerge for this case, as long as we keep the whole filter `(key1=1 and key2=2) or (key1=4 and key3=6)` in the selection.